### DISCUSSION TITLE
Packaging for wasm apps

### DISCUSSION THREAD

**Discussion by siddhantk232** [fastn-stack/fastn#1961] [Ideas & RFCs]
We currently support handling http request via wasm but there's no answer on how to ship these compiled wasm files with a fastn package.

Moreover a wasm package may also want to deliver UI that should be customisable (or overwritable) by the package user.

### Possible use cases

- db explorer package. A wasm package that can be installed in dev mode to store/exec sql queries (like https://pypi.org/project/django-sql-explorer/)
- Our [subscription app](https://github.com/fastn-community/subscription/). We want to have an interface that would aid sending non-transactional emails to subscribers (from html template and even ftd templates[^1] in the future).

[^1]: This requires that a subset of our new ftd language should be compilable to other formats (html/pdfs etc).

**Comment by amitu** [comment]
Generally this is handled by fastn.app. `fastn.app` allows you to "mount" another package as an "app" in the current / main package. App is partially implemented also. All the URLs in the mounted app will be available on the mount url for that app. An app can be mounted more than once also if you wanted more than one instance of the app.

WASM file for a fastn package should be distributed with the package itself, so if I use auth package a dependency, and it needs wasm file, it will be part of auth package, and I will simply mount the auth package on say amitu.com/auth/.

Read these discussions: https://github.com/orgs/fastn-stack/discussions?discussions_q=fastn.app

**Comment by siddhantk232** [comment]
> WASM file for a fastn package should be distributed with the package itself, so if I use auth package a dependency, and it needs wasm file, it will be part of auth package, and I will simply mount the auth package on say amitu.com/auth/.

We loose a lot of information about the compiled wasm binary when we choose to do this. Where the binary is coming from? How do I get dependabot alerts? The packaging can't just be "upload a binary with your package and specify a path to it in our fastn.url-mappings" for these reasons. We need some more metadata (where is the source at least).

**Reply by amitu** [reply]
This information has to live with your Rust code. There is no reason to checkin Rust code along with the fastn package. fastn package should be hosted on fifthtry and rust code on Github. On github to publish, instead of publishing a release, or along with it, we should use fastn upload (the single file variant) to make the latest wasm part of the fastn package.
